filesystems/drs/src/main/scala/cromwell/filesystems/drs/DrsPathBuilderFactory.scala

```diff
@@ -39,37 +41,55 @@ class DrsPathBuilderFactory(globalConfig: Config, instanceConfig: Config, single
   private val GcsScheme = "gs"
 
 
-  private def gcsInputStream(gcsFile: GcsFilePath, serviceAccount: String): IO[ReadableByteChannel] = {
-    val credentials = GoogleCredentials.fromStream(new ByteArrayInputStream(serviceAccount.getBytes()))
+  private def gcsInputStream(gcsFile: GcsFilePath,
+                             serviceAccountJson: String,
+                             requesterPaysProjectIdOption: Option[String]): IO[ReadableByteChannel] = {
+    val credentials = GoogleCredentials.fromStream(new ByteArrayInputStream(serviceAccountJson.getBytes()))
     val storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService
 
     IO.delay {
       val blob = storage.get(gcsFile.bucket, gcsFile.file)
       blob.reader()
+    } handleErrorWith { throwable =>
+      (requesterPaysProjectIdOption, throwable) match {
+        case (Some(requesterPaysProjectId), storageException: StorageException)
+          if storageException.getMessage == "Bucket is requester pays bucket but no user project provided." =>
+          IO.delay {
+            val blob = storage.get(gcsFile.bucket, gcsFile.file, BlobGetOption.userProject(requesterPaysProjectId))
+            blob.reader(Blob.BlobSourceOption.userProject(requesterPaysProjectId))
+          }
+        case _ => IO.raiseError(throwable)
+      }
     }
   }
 
 
-  private def inputReadChannel(url: String, urlScheme: String, serviceAccount: String): IO[ReadableByteChannel] =  {
+  private def inputReadChannel(url: String,
+                               urlScheme: String,
+                               serviceAccountJson: String,
+                               requesterPaysProjectIdOption: Option[String]): IO[ReadableByteChannel] =  {
     urlScheme match {
       case GcsScheme =>
         val Array(bucket, fileToBeLocalized) = url.replace(s"$GcsScheme://", "").split("/", 2)
-        gcsInputStream(GcsFilePath(bucket, fileToBeLocalized), serviceAccount)
+        gcsInputStream(GcsFilePath(bucket, fileToBeLocalized), serviceAccountJson, requesterPaysProjectIdOption)
       case otherScheme => IO.raiseError(new UnsupportedOperationException(s"DRS currently doesn't support reading files for $otherScheme."))
     }
   }
 
 
-  private def drsReadInterpreter(marthaResponse: MarthaResponse): IO[ReadableByteChannel] = {
-    val serviceAccount = marthaResponse.googleServiceAccount match {
-      case Some(googleSA) => googleSA.data.toString
-      case None => throw new GoogleSANotFoundException
+  private def drsReadInterpreter(options: WorkflowOptions, requesterPaysProjectIdOption: Option[String])
+                                (marthaResponse: MarthaResponse): IO[ReadableByteChannel] = {
+    val serviceAccountJsonIo = marthaResponse.googleServiceAccount match {
+      case Some(googleSA) => IO.pure(googleSA.data.toString)
+      case None => IO.fromEither(options.get(GoogleAuthMode.UserServiceAccountKey).toEither)
```

Comment:
Shouldn't this be whatever service account Cromwell is using for the rest of the workflow?

Eg say a user has access to some file but the "Cromwell" account doesn't - that user wouldn't be able to read that file via DRS.

Eg2 say a user does *not* have access to some file which the Cromwell SA does - that user would inappropriately be able to access that file via a DRS indirection

Reply:
The ticket currently says to use the User SA in this case, and fail otherwise. I think this is because the Requester Pays feature is currently needed to support DataSTAGE only, and hence the simple requirements.

Reply:
Exactly